src/vsr/replica.zig

```diff
@@ -370,33 +350,62 @@ pub fn ReplicaType(
             // Open the (Forest inside) StateMachine:
             self.opened = false;
             self.state_machine.open(state_machine_open_callback);
-            while (!self.opened) {
-                // self.grid.tick();
-                self.superblock.storage.tick();
-            }
+            while (!self.opened) self.superblock.storage.tick();
 
             self.opened = false;
             self.journal.recover(journal_recover_callback);
             while (!self.opened) self.superblock.storage.tick();
 
-            if (self.journal.is_empty()) {
-                // The data file is brand new — no messages have ever been written.
-                // Transition to normal status; no need to run the VSR recovery protocol.
-                assert(self.journal.dirty.count == 0);
-                assert(self.journal.faulty.count == 0);
-                assert(self.commit_min == 0);
-                assert(self.commit_max == 0);
-                assert(self.op_checkpoint == 0);
-                assert(self.op == 0);
-                assert(self.view == 0);
-
-                log.debug("{}: open: empty data file", .{self.replica});
-                self.transition_to_normal_from_recovering_status(0);
-                assert(self.status == .normal);
-            } else if (self.replica_count == 1) {
-                if (self.journal.faulty.count != 0) @panic("journal is corrupt");
+            const vsr_headers = self.superblock.working.vsr_headers();
+            var op_head: u64 = vsr_headers.slice[0].op;
+            for (self.journal.headers) |*header| {
+                if (header.command == .prepare and header.op > op_head) {
+                    assert(header.view <= self.log_view);
+                    op_head = header.op;
+                }
+            }
+
+            self.op = op_head;
+            for (vsr_headers.slice) |*header| self.replace_header(header);
```

Comment:
Why do we need to replace on open? Do we not ensure that the journal matches before writing the superblock?

Reply:
> Do we not ensure that the journal matches before writing the superblock?

No — the WAL can lag behind the in-memory journal, but the in-memory journal is lost during a crash.

The in-memory journal is recovered from the WAL at startup, but it isn't necessarily accurate. In particular, the journal is only updated when we write prepares, but the journal is updated with headers from view changes. So we may receive some headers in a `start_view`, but not have those in our WAL. Or there may be entries in our WAL which were truncated from the journal, which "leak through" during recovery — we try to remove these when we can identify them.

Of all of the headers replaced here (in this loop), only the first one is critical, since it might be our new head op, which _must_ exist in the in-memory journal headers. ("head op exists in journal headers" is a replica invariant.)

Reply:
> Of all of the headers replaced here (in this loop), only the first one is critical

All of the headers after that which we replace are just bonus — replacing them here can reduce the number of round-trips required for repair.